include/wex/Engine.hpp

```diff
@@ -0,0 +1,102 @@
+#pragma once
+#include <SFML/Graphics.hpp>
+#include <SFML/Graphics/RenderWindow.hpp>
+#include <SFML/System/Vector2.hpp>
+#include <SFML/Window/Event.hpp>
+#include <iostream>
+#include <memory>
+#include <string>
+
+#include "Common.hpp"
+#include "GraphicsController.hpp"
+#include "util.hpp"
+
+
+namespace wex {
+
+class Game {
+	friend Engine;
+
+  public:
+	/// \brief Intialize the game state. This function is called once
+	/// before the core game loop is started. By default, it does nothing
+	/// and needs to be overriden by the user for providing meaningful
+	/// behavior.
+	virtual void init() {}
+
+	/// \brief The update function runs several times a second and is called in a way
+	/// such that it's independ of the frame rate. This function is ideal for updating
+	/// the state of the game every tick.
+	virtual void update(double dt) = 0;
+
+	/// \brief The draw function is resposible for rendering everything that is visible
+	/// on the screen.
+	/// \param graphics A Graphics Controller that belongs to the engine. the controller contains
+	///	       member functions that can help draw shapes, lines and images to the screen.
+	virtual void draw(GraphicsController& graphics) = 0;
+
+	virtual ~Game() = default;
+
+	inline void setGraphicsController(GraphicsController& controller) noexcept {
+		this->g = std::shared_ptr<GraphicsController>(&controller);
+	}
+
+	inline const std::shared_ptr<GraphicsController> graphicsController() const noexcept {
+		return this->g;
+	}
+
+  protected:
+	std::shared_ptr<GraphicsController> g = nullptr;
+};
+
+struct Config {
+	WindowConfig window;
+	Config() = default;
+};
+
+class Engine final : util::Pinned {
+
+	/// \brief Exit code for the game engine.
+	enum class Status { ok };
+
+  public:
+	~Engine() = default;
+
+	explicit Engine(std::unique_ptr<Game>&& game, Config const& config = Config())
+		: mConfig(std::move(config)), mGame(std::move(game)) {
+		mGame->setGraphicsController(mGraphics);
+	}
+
+	/// \brief Intialize the window and start the game loop.
+	/// \return The status code for the game. Returns Status::ok if the game starts and quits
+	///			successfully.
+	Status run();
+
+  private:
+	/// \brief The core game loop. calls `game.update` and `game.draw`
+	void loop();
+
+	/// \brief polls events from the window until none are left and then
+	/// registers those events.
+	void handleEvents(sf::RenderWindow& window);
```

Comment:
There are sf types in the engine interface, but SFML headers are privately included, so this same header will not compile when included in user code